plutus-core/plutus-ir/src/PlutusIR/Transform/Inline.hs

```diff
@@ -144,29 +183,38 @@ processTerm
+processTerm = handleTerm <=< traverseOf termSubtypes applySubstitution where
+    handleTerm = \case
+        v@(Var _ n) -> do
+            subst <- get
+            pure $ case lookupTerm n subst of
+                -- Not substituted for, leave it as it is
+                Nothing       -> v
+                -- Already processed term, just put it in, don't do any further optimization here.
+                -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
+                -- NOTE:  We need to rename it to avoid variable capture
+                Just (Done t) -> t
+        Let a NonRec bs t -> do
+            -- Process bindings, eliminating those which will be inlined unconditionally,
+            -- and accumulating the new substitutions
+            -- See Note [Removing inlined bindings]
+            -- Note that we don't *remove* the bindings or scope the state, so the state will carry over
+            -- into "sibling" terms. This is fine because we have global uniqueness
+            -- (see Note [Inlining and global uniqueness]), if somewhat wasteful.
+            bs' <- wither processSingleBinding (toList bs)
+            t' <- processTerm t
+            -- Use 'mkLet': we're using lists of bindings rather than NonEmpty since we might actually
+            -- have got rid of all of them!
+            pure $ mkLet a NonRec bs' t'
+        -- This includes recursive let terms, we don't even consider inlining them at the moment
+        t -> forMOf termSubterms t processTerm
+    applySubstitution = typeSubstTyNamesM substFreshenedType
+    substFreshenedType name = do
         subst <- get
-        pure $ case lookupSubst n subst of
-            -- Not substituted for, leave it as it is
-            Nothing       -> v
-            -- Already processed term, just put it in, don't do any further optimization here.
-            -- See Note [Inlining approach and 'Secrets of the GHC Inliner']
-            Just (Done t) -> t
-    Let a NonRec bs t -> do
-        -- Process bindings, eliminating those which will be inlined unconditionally,
-        -- and accumulating the new substitutions
-        -- See Note [Removing inlined bindings]
-        -- Note that we don't *remove* the bindings or scope the state, so the state will carry over
-        -- into "sibling" terms. This is fine because we have global uniqueness
-        -- (see Note [Inlining and global uniqueness]), if somewhat wasteful.
-        bs' <- wither processSingleBinding (toList bs)
-        t' <- processTerm t
-        -- Use 'mkLet': we're using lists of bindings rather than NonEmpty since we might actually
-        -- have got rid of all of them!
-        pure $ mkLet a NonRec bs' t'
-    -- This includes recursive let terms, we don't even consider inlining them at the moment
-    t -> forMOf termSubterms t processTerm
+        case lookupType name subst of
+            Nothing -> pure Nothing
+            Just t  -> PLC.rename t <&> Just
```

Comment:
Probably worth a Note about our renaming strategy too. We discussed it quite a bit - let's make sure we record that!